checker/src/features/narrowing.rs

```diff
@@ -0,0 +1,470 @@
+					let narrowed_to =
+						types.new_intrinsic(&crate::types::intrinsics::Intrinsic::LessThan, rhs);
+					let narrowed = types.new_narrowed(lhs, narrowed_to);
+					into.insert(lhs, narrowed);
+				} else if types.get_type_by_id(rhs).is_dependent() {
+					let narrowed_to =
+						types.new_intrinsic(&crate::types::intrinsics::Intrinsic::GreaterThan, lhs);
+					let narrowed = types.new_narrowed(rhs, narrowed_to);
+					into.insert(rhs, narrowed);
+				}
+			}
+			Constructor::UnaryOperator { operator: PureUnary::LogicalNot, operand } => {
+				narrow_based_on_expression(*operand, !negate, into, information, types);
+			}
+			Constructor::TypeOperator(TypeOperator::IsPrototype { lhs, rhs_prototype }) => {
+				let (lhs, rhs_prototype) = (*lhs, *rhs_prototype);
+				let constraint = crate::types::get_constraint(lhs, types).unwrap_or(lhs);
+				// TODO want a mix of two
+				let narrowed_to = if !negate && constraint == TypeId::ANY_TYPE {
+					generate_new_type_with_prototype(rhs_prototype, types)
+				} else {
+					let mut result = Vec::new();
+					let filter = Filter::HasPrototype(rhs_prototype);
+					let filter = if negate { Filter::Not(&filter) } else { filter };
+					build_union_from_filter(constraint, filter, &mut result, information, types);
+					types.new_or_type_from_iterator(result)
+				};
+				let narrowed = types.new_narrowed(lhs, narrowed_to);
+				into.insert(lhs, narrowed);
+			}
+			Constructor::TypeExtends(crate::types::TypeExtends { item, extends }) => {
+				let (item, extends) = (*item, *extends);
+				// experimental
+				// TODO want a mix of two
+				let constraint = crate::types::get_constraint(item, types).unwrap_or(item);
+				let narrowed_to = if !negate && constraint == TypeId::ANY_TYPE {
+					extends
+				} else {
+					let mut result = Vec::new();
+					let filter = Filter::IsType(extends);
+					let filter = if negate { Filter::Not(&filter) } else { filter };
+					build_union_from_filter(constraint, filter, &mut result, information, types);
+					types.new_or_type_from_iterator(result)
+				};
+				let narrowed = types.new_narrowed(item, narrowed_to);
+				into.insert(item, narrowed);
+			}
+			Constructor::TypeOperator(TypeOperator::HasProperty(on, under)) => {
+				let on = *on;
+				let constraint = types::get_constraint(on, types).unwrap_or(on);
+				let narrowed_to = if !negate && constraint == TypeId::ANY_TYPE {
+					generate_new_type_with_property(under.into_owned(), TypeId::ANY_TYPE, types)
+				} else {
+					let mut items = Vec::new();
+					build_union_from_filter(
+						constraint,
+						Filter::HasProperty {
+							property: under,
+							filter: &Filter::IsType(TypeId::ANY_TYPE),
+						},
+						&mut items,
+						information,
+						types,
+					);
+					types.new_or_type_from_iterator(items)
+				};
+				into.insert(on, types.new_narrowed(on, narrowed_to));
+			}
+			constructor => {
+				if let Some((lhs, rhs)) = as_logical_and(constructor, types) {
```

Comment:
@danvk thought you might find this interesting! I watched the [Michigan TypeScript](https://www.youtube.com/@MichiganTypeScript) video and read your blog post on [implementing inferred type predicates](https://effectivetypescript.com/2024/04/16/inferring-a-type-predicate/). (congrats on getting it merged, it is a great feature!)

In Ezno it was a lot easier to add narrowing with its more *dependent* type system. In Ezno the expression `typeof x === "number"` has a type like `EqOperator<TypeOfOperator<typeof x>, "number">` which makes getting the information for narrowing pretty simple (you can sort of see that in the lines above). So Ezno has something closer to your original intuition rather than the TSC flow node control flow structure.

One of the interesting parts was "Pathological Cases and an Insight" with this case edge case in the `else`
```ts
function flakyIsStringRewrite(x: string | null) {
  if (typeof x === 'string' && Math.random() > 0.5) {
    x; // type is string
  } else {
    x; // type is string | null
  }
}
```
this helped me fix a bug in my implementation. For else branches I run `get_narrowed_values(condition, true)`, where `true` negates the condition in the `if`. If I have the `&&` expression above with `negate = true` it then means that it applies one of the De Morgans law. So it now treats `!(a && b)` in the code as `!a || !b` and it only adds a *narrow* if both sides have a condition for the same variable (=`on` here) (the zipping code being `for (on, lhs_request) in lhs_requests {	if let Some(rhs_request) = rhs_requests.get(&on) { ...`). 

[I think I have got this covered in the tests here](https://github.com/kaleidawave/ezno/blob/narrowing/checker/specification/staging.md#edge-case)

Don't know if there is an equivalent section in the TS narrowing logic but thought my intuition thinking about De Morgan's laws and generating `A | B` only if both sides of an `||` condition give narrowed values for the variable was a nice way to handle this case.